diff --git a/MCStructs-text/src/main/java/net/lenni0451/mcstructs/text/TextComponent.java b/MCStructs-text/src/main/java/net/lenni0451/mcstructs/text/TextComponent.java
--- a/MCStructs-text/src/main/java/net/lenni0451/mcstructs/text/TextComponent.java	(revision 4f60ce806befcb04cf9ef8984f6d892c52034f62)
+++ b/MCStructs-text/src/main/java/net/lenni0451/mcstructs/text/TextComponent.java	(date 1736614412299)
@@ -36,6 +36,13 @@
 
     private final List<TextComponent> siblings = new ArrayList<>();
     private Style style = new Style();
+    private int level = 1;
+    private void setLevel(int level) {
+        this.level = Math.max(this.level, level);
+        if (this.level > 512) {
+            throw new IllegalArgumentException("Too deep");
+        }
+    }
 
     /**
      * Append multiple strings to this component.
@@ -56,6 +63,7 @@
      */
     public TextComponent append(final TextComponent component) {
         this.siblings.add(component);
+        component.setLevel(this.level + 1);
         return this;
     }
 
@@ -67,11 +75,13 @@
      */
     public TextComponent append(final TextComponent... components) {
         Collections.addAll(this.siblings, components);
+        for (TextComponent component : components) component.setLevel(this.level + 1);
         return this;
     }
 
     public TextComponent append(final Iterable<TextComponent> components) {
         components.forEach(this.siblings::add);
+        for (TextComponent component : components) component.setLevel(this.level + 1);
         return this;
     }

diff --git a/MCStructs-text/src/test/java/net/lenni0451/mcstructs/text/serializer/TextComponentCodecTest.java b/MCStructs-text/src/test/java/net/lenni0451/mcstructs/text/serializer/TextComponentCodecTest.java
--- a/MCStructs-text/src/test/java/net/lenni0451/mcstructs/text/serializer/TextComponentCodecTest.java	(revision 4f60ce806befcb04cf9ef8984f6d892c52034f62)
+++ b/MCStructs-text/src/test/java/net/lenni0451/mcstructs/text/serializer/TextComponentCodecTest.java	(date 1736615658895)
@@ -58,9 +58,9 @@
 
     @Test
     void legacyItemDeserialization() throws SNbtSerializeException {
-        CompoundTag legacyNbt = new CompoundTag()
-                .put("id", "stone")
-                .putByte("Count", (byte) 5);
+        CompoundTag legacyNbt = new CompoundTag();
+        legacyNbt.put("id", new StringTag("stone"));
+        legacyNbt.putByte("Count", (byte) 5);
         TextComponent legacyComponent = new StringComponent("test")
                 .setStyle(new Style()
                         .setHoverEvent(new TextHoverEvent(HoverEventAction.SHOW_ITEM, new StringComponent(SNbt.LATEST.serialize(legacyNbt))))
@@ -77,10 +77,10 @@
     @Test
     void legacyEntityDeserialization() throws SNbtSerializeException {
         UUID randomUUID = UUID.randomUUID();
-        CompoundTag legacyNbt = new CompoundTag()
-                .put("name", "{\"text\":\"test\"}")
-                .put("type", "cow")
-                .put("id", randomUUID.toString());
+        CompoundTag legacyNbt = new CompoundTag();
+        legacyNbt.put("name", new StringTag("{\"text\":\"test\"}"));
+        legacyNbt.put("type", new StringTag("cow"));
+        legacyNbt.put("id", new StringTag(randomUUID.toString()));
         TextComponent legacyComponent = new StringComponent("test")
                 .setStyle(new Style()
                         .setHoverEvent(new TextHoverEvent(HoverEventAction.SHOW_ENTITY, new StringComponent(SNbt.LATEST.serialize(legacyNbt))))
@@ -97,19 +97,31 @@
 
     @Test
     void arrayWithTag() {
-        ListTag<Tag> tags = new ListTag<>()
-                .add(new CompoundTag()
-                        .putString("translate", "test")
-                        .addByteArray("with", (byte) 1, (byte) 2, (byte) 3))
-                .add(new CompoundTag()
-                        .putString("translate", "test")
-                        .addIntArray("with", 1, 2, 3))
-                .add(new CompoundTag()
-                        .putString("translate", "test")
-                        .addLongArray("with", 1, 2, 3))
-                .add(new CompoundTag()
-                        .putString("translate", "test")
-                        .addList("with", 1, 2, 3));
+        CompoundTag translateWithByteArray = new CompoundTag();
+        translateWithByteArray.putString("translate", "test");
+        translateWithByteArray.put("with", new ByteArrayTag(new byte[]{1, 2, 3}));
+
+        CompoundTag translateWithIntArray = new CompoundTag();
+        translateWithIntArray.putString("translate", "test");
+        translateWithIntArray.put("with", new IntArrayTag(new int[]{1, 2, 3}));
+
+        CompoundTag translateWithLongArray = new CompoundTag();
+        translateWithLongArray.putString("translate", "test");
+        translateWithLongArray.put("with", new LongArrayTag(new long[]{1, 2, 3}));
+
+        CompoundTag translateWithList = new CompoundTag();
+        ListTag<IntTag> numberList = new ListTag<>(IntTag.class);
+        numberList.add(new IntTag(1));
+        numberList.add(new IntTag(2));
+        numberList.add(new IntTag(3));
+        translateWithList.putString("translate", "test");
+        translateWithList.put("with", numberList);
+
+        ListTag<CompoundTag> tags = new ListTag<>(CompoundTag.class);
+        tags.add(translateWithByteArray);
+        tags.add(translateWithIntArray);
+        tags.add(translateWithLongArray);
+        tags.add(translateWithList);
         TextComponent component = new TranslationComponent("test", (byte) 1, (byte) 2, (byte) 3)
                 .append(new TranslationComponent("test", 1, 2, 3))
                 .append(new TranslationComponent("test", 1L, 2L, 3L))

diff --git a/MCStructs-text/src/main/java/net/lenni0451/mcstructs/text/utils/JsonNbtConverter.java b/MCStructs-text/src/main/java/net/lenni0451/mcstructs/text/utils/JsonNbtConverter.java
--- a/MCStructs-text/src/main/java/net/lenni0451/mcstructs/text/utils/JsonNbtConverter.java	(revision 4f60ce806befcb04cf9ef8984f6d892c52034f62)
+++ b/MCStructs-text/src/main/java/net/lenni0451/mcstructs/text/utils/JsonNbtConverter.java	(date 1736615590939)
@@ -25,23 +25,16 @@
     @Nullable
     public static JsonElement toJson(@Nullable final Tag tag) {
         if (tag == null) return null;
-        switch (tag) {
-            case END:
-                return null;
-            case BYTE:
-            case SHORT:
-            case INT:
-            case LONG:
-            case FLOAT:
-            case DOUBLE:
+
+        if (tag instanceof NumberTag) {
                 return new JsonPrimitive(((NumberTag) tag).getValue());
-            case BYTE_ARRAY:
+        } else if (tag instanceof ByteArrayTag) {
                 JsonArray byteArray = new JsonArray();
                 for (byte b : ((ByteArrayTag) tag).getValue()) byteArray.add(b);
                 return byteArray;
-            case STRING:
+        } else if (tag instanceof StringTag) {
                 return new JsonPrimitive(((StringTag) tag).getValue());
-            case LIST:
+        } else if (tag instanceof ListTag<?>) {
                 JsonArray list = new JsonArray();
                 ListTag<Tag> listTag = ((ListTag) tag);
                 for (Tag tagInList : listTag.getValue()) {
@@ -55,19 +48,19 @@
                     list.add(toJson(tagInList));
                 }
                 return list;
-            case COMPOUND:
+        } else if (tag instanceof CompoundTag) {
                 JsonObject compound = new JsonObject();
-                for (Map.Entry<String, Tag> entry : ((CompoundTag) tag).getValue().entrySet()) compound.put(entry.getKey(), toJson(entry.getValue()));
+                for (Map.Entry<String, Tag> entry : ((CompoundTag) tag).getValue().entrySet()) compound.add(entry.getKey(), toJson(entry.getValue()));
                 return compound;
-            case INT_ARRAY:
+        } else if (tag instanceof IntArrayTag) {
                 JsonArray intArray = new JsonArray();
                 for (int i : ((IntArrayTag) tag).getValue()) intArray.add(i);
                 return intArray;
-            case LONG_ARRAY:
+        } else if (tag instanceof LongArrayTag) {
                 JsonArray longArray = new JsonArray();
                 for (long l : ((LongArrayTag) tag).getValue()) longArray.add(l);
                 return longArray;
-            default:
+        } else {
                 throw new IllegalArgumentException("Unknown Nbt type: " + tag);
         }
     }
@@ -90,31 +83,37 @@
             JsonArray array = element.getAsJsonArray();
             List<Tag> nbtTags = new ArrayList<>();
             Tag listType = null;
+            boolean mixedList = false;
             for (JsonElement arrayElement : array) {
                 Tag tag = toNbt(arrayElement);
                 nbtTags.add(tag);
                 listType = getListType(listType, tag);
+                if (listType == null) mixedList = true;
             }
             if (listType == null) {
                 return new ListTag<>();
-            } else if (listType == Tag.END) { //Mixed list
+            } else if (mixedList) { //Mixed list
                 ListTag<CompoundTag> list = new ListTag<>();
                 for (Tag tag : nbtTags) {
                     if (tag instanceof CompoundTag) list.add(((CompoundTag) tag));
-                    else list.add(new CompoundTag().put("", tag));
+                    else {
+                        final CompoundTag entries = new CompoundTag();
+                        entries.put("", tag);
+                        list.add(entries);
+                    }
                 }
                 return list;
-            } else if (listType == Tag.BYTE) {
+            } else if (listType instanceof ByteTag) {
                 byte[] bytes = new byte[nbtTags.size()];
-                for (int i = 0; i < nbtTags.size(); i++) bytes[i] = nbtTags.get(i).asByteTag().byteValue();
+                for (int i = 0; i < nbtTags.size(); i++) bytes[i] = ((NumberTag) nbtTags.get(i)).asByte();
                 return new ByteArrayTag(bytes);
-            } else if (listType == Tag.INT) {
+            } else if (listType instanceof IntTag) {
                 int[] ints = new int[nbtTags.size()];
-                for (int i = 0; i < nbtTags.size(); i++) ints[i] = nbtTags.get(i).asIntTag().asInt();
+                for (int i = 0; i < nbtTags.size(); i++) ints[i] = ((NumberTag) nbtTags.get(i)).asInt();
                 return new IntArrayTag(ints);
-            } else if (listType == Tag.LONG) {
+            } else if (listType instanceof LongTag) {
                 long[] longs = new long[nbtTags.size()];
-                for (int i = 0; i < nbtTags.size(); i++) longs[i] = nbtTags.get(i).asIntTag().asInt();
+                for (int i = 0; i < nbtTags.size(); i++) longs[i] = ((NumberTag) nbtTags.get(i)).asLong();
                 return new LongArrayTag(longs);
             } else {
                 return new ListTag<>(nbtTags);
@@ -148,7 +147,7 @@
 
     private static Tag getListType(final Tag current, final Tag tag) {
         if (current == null) return tag;
-        if (current != tag) return Tag.END; //Placeholder for mixed lists
+        if (tag == null || current.getClass() != tag.getClass()) return null; //Placeholder for mixed lists
         return current;
     }

diff --git a/MCStructs-text/build.gradle b/MCStructs-text/build.gradle
--- a/MCStructs-text/build.gradle	(revision 4f60ce806befcb04cf9ef8984f6d892c52034f62)
+++ b/MCStructs-text/build.gradle	(date 1736615884164)
@@ -1,7 +1,6 @@
 dependencies {
     api "com.google.code.gson:gson:2.11.0"
 
-    api project(":MCStructs-converter")
     api project(":MCStructs-core")
     api "com.viaversion:nbt:5.0.0"
     api project(":MCStructs-snbt")

diff --git a/build.gradle b/build.gradle
--- a/build.gradle	(revision 4f60ce806befcb04cf9ef8984f6d892c52034f62)
+++ b/build.gradle	(date 1736615884157)
@@ -69,17 +69,17 @@
     publishing {
         repositories {
             maven {
-                name = "reposilite"
-                def releasesUrl = "https://maven.lenni0451.net/releases"
-                def snapshotsUrl = "https://maven.lenni0451.net/snapshots"
-                url = project.maven_version.endsWith("SNAPSHOT") ? snapshotsUrl : releasesUrl
-
-                credentials(PasswordCredentials)
+                name = "Via"
+                url = uri("https://repo.viaversion.com/")
+                credentials {
+                    username = System.getenv("via_username")
+                    password = System.getenv("via_password")
+                }
                 authentication {
-                    basic(BasicAuthentication)
+                    create(BasicAuthentication)
                 }
             }
-            maven {
+            /*maven {
                 name = "ossrh"
                 def releasesUrl = "https://s01.oss.sonatype.org/service/local/staging/deploy/maven2/"
                 def snapshotsUrl = "https://s01.oss.sonatype.org/content/repositories/snapshots/"
@@ -89,7 +89,7 @@
                 authentication {
                     basic(BasicAuthentication)
                 }
-            }
+            }*/
         }
         publications {
             maven(MavenPublication) {

diff --git a/MCStructs-text/src/main/java/net/lenni0451/mcstructs/text/stringformat/StringFormat.java b/MCStructs-text/src/main/java/net/lenni0451/mcstructs/text/stringformat/StringFormat.java
--- a/MCStructs-text/src/main/java/net/lenni0451/mcstructs/text/stringformat/StringFormat.java	(revision 4f60ce806befcb04cf9ef8984f6d892c52034f62)
+++ b/MCStructs-text/src/main/java/net/lenni0451/mcstructs/text/stringformat/StringFormat.java	(date 1736615941914)
@@ -195,10 +195,16 @@
         Style lastStyle = null;
         List<TextComponent> texts = new ArrayList<>();
         texts.add(component.copy().mergeSiblingParentStyle());
+        int translations = 0;
         while (!texts.isEmpty()) {
             TextComponent current = texts.remove(0);
             if (current instanceof TranslationComponent) {
                 TranslationComponent translation = (TranslationComponent) current;
+                TextComponent resolved = translation.resolveIntoComponents();
+                translations += resolved.getSiblings().size();
+                if (translations >= 100) {
+                    throw new IllegalStateException("Too many translations in a single component");
+                }
                 texts.add(0, translation.resolveIntoComponents().mergeSiblingParentStyle());
                 for (int i = 0; i < translation.getSiblings().size(); i++) {
                     texts.add(i + 1, translation.getSiblings().get(i));

diff --git a/MCStructs-snbt/src/main/java/net/lenni0451/mcstructs/snbt/impl/v1_12/SNbtDeserializer_v1_12.java b/MCStructs-snbt/src/main/java/net/lenni0451/mcstructs/snbt/impl/v1_12/SNbtDeserializer_v1_12.java
--- a/MCStructs-snbt/src/main/java/net/lenni0451/mcstructs/snbt/impl/v1_12/SNbtDeserializer_v1_12.java	(revision 4f60ce806befcb04cf9ef8984f6d892c52034f62)
+++ b/MCStructs-snbt/src/main/java/net/lenni0451/mcstructs/snbt/impl/v1_12/SNbtDeserializer_v1_12.java	(date 1736614294910)
@@ -95,10 +95,28 @@
         reader.read();
         reader.skipWhitespaces();
         if (!reader.canRead()) throw this.makeException(reader, "Expected value");
-        else if (c == 'B') return new ByteArrayTag(this.readPrimitiveList(reader, ByteTag.class, ByteArrayTag.class));
-        else if (c == 'L') return new LongArrayTag(this.readPrimitiveList(reader, LongTag.class, LongArrayTag.class));
-        else if (c == 'I') return new IntArrayTag(this.readPrimitiveList(reader, IntTag.class, IntArrayTag.class));
-        else throw new SNbtDeserializeException("Invalid array type '" + c + "' found");
+        else if (c == 'B') {
+            final ListTag<ByteTag> tags = this.readPrimitiveList(reader, ByteTag.class, ByteArrayTag.class);
+            final byte[] array = new byte[tags.size()];
+            for (int i = 0; i < tags.size(); i++) {
+                array[i] = tags.get(i).asByte();
+            }
+            return new ByteArrayTag(array);
+        } else if (c == 'L') {
+            final ListTag<LongTag> tags = this.readPrimitiveList(reader, LongTag.class, LongArrayTag.class);
+            final long[] array = new long[tags.size()];
+            for (int i = 0; i < tags.size(); i++) {
+                array[i] = tags.get(i).asLong();
+            }
+            return new LongArrayTag(array);
+        } else if (c == 'I') {
+            final ListTag<IntTag> tags = this.readPrimitiveList(reader, IntTag.class, IntArrayTag.class);
+            final int[] array = new int[tags.size()];
+            for (int i = 0; i < tags.size(); i++) {
+                array[i] = tags.get(i).asInt();
+            }
+            return new IntArrayTag(array);
+        } else throw new SNbtDeserializeException("Invalid array type '" + c + "' found");
     }
 
     protected Tag readValue(final StringReader_v1_12 reader) throws SNbtDeserializeException {

diff --git a/settings.gradle b/settings.gradle
--- a/settings.gradle	(revision 4f60ce806befcb04cf9ef8984f6d892c52034f62)
+++ b/settings.gradle	(date 1736615884172)
@@ -11,11 +11,6 @@
 
 rootProject.name = "MCStructs"
 
-include(":MCStructs-all")
-include(":MCStructs-converter")
 include(":MCStructs-core")
-include(":MCStructs-data")
-include(":MCStructs-itemcomponents")
-include(":MCStructs-nbt")
 include(":MCStructs-snbt")
 include(":MCStructs-text")

diff --git a/MCStructs-text/src/main/java/net/lenni0451/mcstructs/text/events/hover/impl/EntityHoverEvent.java b/MCStructs-text/src/main/java/net/lenni0451/mcstructs/text/events/hover/impl/EntityHoverEvent.java
--- a/MCStructs-text/src/main/java/net/lenni0451/mcstructs/text/events/hover/impl/EntityHoverEvent.java	(revision 4f60ce806befcb04cf9ef8984f6d892c52034f62)
+++ b/MCStructs-text/src/main/java/net/lenni0451/mcstructs/text/events/hover/impl/EntityHoverEvent.java	(date 1736615345426)
@@ -121,10 +121,10 @@
     @Override
     public String toString() {
         return ToString.of(this)
-                .put("action", this.action)
-                .put("entityType", this.entityType)
-                .put("uuid", this.uuid)
-                .put("name", this.name, Objects::nonNull)
+                .add("action", this.action)
+                .add("entityType", this.entityType)
+                .add("uuid", this.uuid)
+                .add("name", this.name, Objects::nonNull)
                 .toString();
     }

diff --git a/MCStructs-text/src/main/java/net/lenni0451/mcstructs/text/events/hover/impl/ItemHoverEvent.java b/MCStructs-text/src/main/java/net/lenni0451/mcstructs/text/events/hover/impl/ItemHoverEvent.java
--- a/MCStructs-text/src/main/java/net/lenni0451/mcstructs/text/events/hover/impl/ItemHoverEvent.java	(revision 4f60ce806befcb04cf9ef8984f6d892c52034f62)
+++ b/MCStructs-text/src/main/java/net/lenni0451/mcstructs/text/events/hover/impl/ItemHoverEvent.java	(date 1736615345433)
@@ -120,10 +120,10 @@
     @Override
     public String toString() {
         return ToString.of(this)
-                .put("action", this.action)
-                .put("item", this.item)
-                .put("count", this.count, count -> count != 1)
-                .put("nbt", this.nbt, Objects::nonNull)
+                .add("action", this.action)
+                .add("item", this.item)
+                .add("count", this.count, count -> count != 1)
+                .add("nbt", this.nbt, Objects::nonNull)
                 .toString();
     }
 
